merlin-server/sql/merlin/functions/hasura/hasura_functions.sql

```diff
@@ -27,3 +27,58 @@ begin
 	  order by p.name, ps.start_offset desc;
 end
 $$;
+
+create function hasura_functions.restore_activity_directive_from_changelog(
+  _plan_id integer,
+  _activity_directive_id integer,
+  _revision integer,
+  hasura_session json)
+  returns setof activity_directive
+  language plpgsql
+  volatile
+as $$
+declare
+  _function_permission metadata.permission;
+begin
+  _function_permission :=
+      metadata.get_function_permissions('restore_activity_directive_from_changelog', hasura_session);
+  if not _function_permission = 'NO_CHECK' then
+    call metadata.check_general_permissions(
+        'restore_activity_directive_from_changelog',
+        _function_permission, _plan_id,
+        (hasura_session ->> 'x-hasura-user-id')
+      );
+  end if;
+
+  if not exists(select revision
+                from public.activity_directive_changelog
+                where (plan_id, activity_directive_id, revision) =
+                      (_plan_id, _activity_directive_id, _revision))
+  then
+    raise exception 'Changelog Revision % does not exist for Plan % and Activity %', _revision, _plan_id, _activity_directive_id;
+  end if;
+
+  update activity_directive
+  set name                       = c.name,
+      source_scheduling_goal_id  = c.source_scheduling_goal_id,
+      start_offset               = c.start_offset,
+      type                       = c.type,
+      arguments                  = c.arguments,
+      last_modified_arguments_at = c.changed_arguments_at,
+      metadata                   = c.metadata,
+      anchor_id                  = c.anchor_id,
+      anchored_to_start          = c.anchored_to_start,
+      last_modified_at           = c.changed_at,
+      last_modified_by           = c.changed_by
+  from activity_directive_changelog as c
+  where id = _activity_directive_id
+    and c.plan_id = _plan_id
+    and c.revision = _revision;
+
+  return query select *
+               from activity_directive
+               where id = _activity_directive_id
+                 and plan_id = _plan_id;
+
```

Comment:
This can be simplified into just:
```sql
return query 
update activity_directive
  set name                       = c.name,
      source_scheduling_goal_id  = c.source_scheduling_goal_id,
      start_offset               = c.start_offset,
      type                       = c.type,
      arguments                  = c.arguments,
      last_modified_arguments_at = c.changed_arguments_at,
      metadata                   = c.metadata,
      anchor_id                  = c.anchor_id,
      anchored_to_start          = c.anchored_to_start,
      last_modified_at           = c.changed_at,
      last_modified_by           = c.changed_by
  from activity_directive_changelog as c
  where id = _activity_directive_id
    and c.plan_id = _plan_id
    and c.revision = _revision
  returning *;

```